stdlib/public/Concurrency/Actor.cpp

```diff
@@ -1498,56 +1389,89 @@ void DefaultActorImpl::enqueueStealer(Job *job, JobPriority priority) {
+  while (job) {
+    auto next = getNextJob(job);
+    result.prepend(job);
+    job = next;
+  }
+
+  if (!result.head) {
+    return result;
+  }
 
+  // We must ensure that any jobs not seen by the collectJobs() don't have any
+  // dangling references to the jobs that have been collected. For that we must
+  // atomically set head pointer to NULL. If it fails because more jobs have
+  // been added in the meantime, we have to continue collecting jobs.
   while (true) {
     assert(oldState.isAnyRunning());
 
-    if (!firstJob) {
-      // Nothing to drain, short circuit
-      SWIFT_TASK_DEBUG_LOG("No jobs to drain on actor %p", this);
-      return NULL;
-    }
-
     auto newState = oldState;
-    // Dequeue the first job and set up a new head
-    newState = newState.withFirstJob(getNextJobInQueue(firstJob));
-    if (_status().compare_exchange_weak(oldState, newState,
-                            /* success */ std::memory_order_relaxed,
-                            /* failure */ std::memory_order_relaxed)) {
-      SWIFT_TASK_DEBUG_LOG("Drained first job %p from actor %p", firstJob, this);
-      traceActorStateTransition(this, oldState, newState, distributedActorIsRemote);
-      concurrency::trace::actor_dequeue(this, firstJob);
-      return firstJob;
+    newState = newState.withFirstUnprioritisedJob(nullptr);
+
+    if (_status().compare_exchange_weak(
+            oldState, newState,
+            /* success */ std::memory_order_relaxed,
+            /* failure */ std::memory_order_relaxed)) {
+      SWIFT_TASK_DEBUG_LOG("Collected some jobs from actor %p", this);
+      traceActorStateTransition(this, oldState, newState,
+                                distributedActorIsRemote);
+      return result;
     }
 
     // We failed the weak cmpxchg spuriously, go through loop again.
-    if (oldState.getFirstJob().getAsJob() == jobToPreprocessFrom.getAsJob()) {
+    if (oldState.getFirstUnprioritisedJob() == result.tail) {
       continue;
     }
 
     // There were new items concurrently added to the queue. We need to
-    // preprocess the newly added unprocessed items and merge them to the already
-    // preprocessed list.
+    // collect the newly added unprocessed items and append them to the
+    // already collected list.
     //
-    // The newly merged items that need to be preprocessed, are between the head
-    // of the linked list, and the last job we did the previous preprocessQueue
-    // on
-    firstJob = preprocessQueue(oldState.getFirstJob(), jobToPreprocessFrom, firstJob);
-    jobToPreprocessFrom = oldState.getFirstJob();
-    traceJobQueue(this, firstJob);
+    job = oldState.getFirstUnprioritisedJob();
+    SimpleQueue extraJobs;
+    while (job != result.tail) {
+      auto next = getNextJob(job);
+      extraJobs.prepend(job);
+      job = next;
+    }
+    result.appendContentsOf(extraJobs);
+  }
+}
+
+// Called with actor lock held on current thread
+void DefaultActorImpl::processJobs() {
```

Comment:
We're not really processing the jobs here.  How about `processIncomingQueue`?  And please add a doc comment, something like:

```
/// Check for new jobs in the incoming queue and move them to the
/// processing queue.
```

Also, while we need to call this *between* jobs just in case a higher-priority job appeared on the queue, I think the current call structure means we'll actually do two updates of `status` when we're just *starting* to process the queue in `defaultActorDrain`: one in `tryLock` in order to claim the actor, and then another in `collectJobs`.  That's necessary in the old approach because everything is tied together: you have to hold the actor lock before you can look at the job queue, and draining a job from the queue requires looking at it, so of course we have to do one exchange to claim the actor lock and then another to preprocess the job queue and drain a job from it.  But in the new approach, there's no reason that basically every update to the actor status that ends with a postcondition of holding the actor lock shouldn't also move incoming jobs to the processing queue.

So I would suggest something like this:
- `handleUnprioritizedJobs(Job *)` takes a claimed incoming job queue and processes it into the processing queue.
- `processIncomingQueue()` checks for new incoming jobs, claims them if present, and calls `handleUnprioritizedJobs`.
- `tryLock` *also* claims any incoming jobs and calls `handleUnprioritizedJobs` if it successfully claims the lock.  (Maybe only if `asDrainer` is true?  There should only be incoming jobs in this case if our job is higher priority than them, and we don't need to eagerly process lower-priority jobs in that case.)
- `drainOne` should not call `processIncomingQueue()`; it should only work with the processing queue.
- `defaultActorDrain` should call `processIncomingQueue()` between jobs, assuming it's decided to continue running.